File: service/internal/infrastructure/health/healthchecker.go
package health

import "context"

// Healthchecker represents any component that can report on its own health.
type Healthchecker interface {
	// Healthcheck will check the health of the component and return an error if the component is unhealthy.
	Healthcheck(ctx context.Context) error
}

// HealthcheckerFunc is a function type that implements the Healthchecker interface.
type HealthcheckerFunc func(context.Context) error

// Healthcheck will check the health of the component and return an error if the component is unhealthy.
func (f HealthcheckerFunc) Healthcheck(ctx context.Context) error {
	return f(ctx)
}

File: service/src/health/healthchecker.rs
use std::{boxed::Box, error::Error};

/// Trait that components able to report on their health can implement.
pub trait Healthchecker {
    /// Check the health of the component.
    ///
    /// # Returns
    /// If the component is healthy then returns `Ok(())`.
    /// If the component is unhealthy then returns `Err()` containing the reason why the component is unhealthy.
    fn check_health(&self) -> Result<(), Box<dyn Error>>;
}
